thunder/dynamo/report.py

```diff
@@ -399,6 +401,200 @@ def write_repro(
+
+        if file_name is None:
+            file_name = f"{self.graph_name}.py"
+        with open(folder / file_name, "w") as f:
+            print(code_str, file=f)
+
+    def run_benchmark(self, compile_fn: CompileSpecificationInterface, time_fn: TimerInterface):
+        compiled_fn = compile_fn.compile(self.graph)
+        inputs = self.get_input_metadata()
+        forward_only = not any(hasattr(arg, "requires_grad") and arg.requires_grad for arg in inputs)
+        example_inputs = eval(f"[\n{chr(10).join(arg_like(a) for a in inputs)}]")
+        fwd_measurement = time_fn.time(
+            "compiled_fn(*example_inputs)", globals={"compiled_fn": compiled_fn, "example_inputs": example_inputs}
+        )
+        if forward_only:
+            return fwd_measurement
+        else:
+            backward_fn, backward_setup = backward_only(compiled_fn, *example_inputs)
+            backward_args = backward_setup()
+            bwd_measurement = time_fn.time(
+                "backward_fn(*backward_args)", globals={"backward_fn": backward_fn, "backward_args": backward_args}
+            )
+            return fwd_measurement, bwd_measurement
+
+    def write_benchmark(
+        self,
+        folder: str | PathLike,
+        compile_fn: CompileSpecificationInterface,
+        time_fn: TimerInterface,
+        file_name: str = None,
+        serialize_inputs: bool = False,
+        inputs: Sequence[torch.Tensor | ExampleInputMetaData] = None,
+        **kwargs,
+    ):
+        """
+        Generates a benchmark reproduction script for the given compilation and timing specification and writes it to the specified file.
+
+        Args:
+            folder (str | PathLike): The target directory where the script will be saved.
+            compile_fn (CompileSpecificationInterface): Specifies how the FX graph module should be compiled.
+                See :class:`CompileSpecificationInterface` for details.
+            time_fn(TimerInterface): Specifies how the compiled callable is timed. See :class:`TimerInterface` for details.
+            file_name (str): The name of the output script file. Default is the :attr:`graph_name`
+            serialize_inputs (bool, optional): Whether to serialize the inputs for reproducibility. Defaults to False.
+                If enabled, all inputs will be serialized into a single file: "{graph_name}_input.pt".
+            inputs (Sequence[torch.Tensor | ExampleInputMetaData], optional):
+                The input tensors or metadata for the FX graph module. Defaults to None.
+                If not provided, inputs will be inferred from the placeholders in the FX graph.
+            **kwargs: Additional arguments for customization.
+        """
+        folder = Path(folder)
+        folder.mkdir(exist_ok=True, parents=True)
+        if inputs == None:
+            inputs = self.get_input_metadata()
+        forward_only = not any(hasattr(arg, "requires_grad") and arg.requires_grad for arg in inputs)
```

Comment:
Seems like there should be an opportunity to reuse more code with `write_repro_v2`